# March 2022

## New Py Scripts

### Notes added to the NST_notes.md file

- apps/toonify/src/NST_train_single_image.py >> Run this file to train NST on a new style and input pair. At least 2000 epochs are recommended.

- apps/toonify/src/stylization_script.py >> Run this file to transfer input image into pre-trained styles. There are 4 pre-trained models are available in models/binaries folder.

- apps/toonify/src/training_script.py >> To train NST on new style. This will save the pre-trained model which can be used to transfer images afterwards.

- apps/toonify/src/utils/ >> This utils folder consists of scripts for download pre-trained models from dropbox and other helper functions.

### Answer

I have a question about these files: NST_train_single_image.py, stylization_script.py and training_script.py

When I used the NST script, I just used a style image and an input image and called NST.py like this:
python apps/toonify/src/NST.py --input_dir input.jpg --style_image style.jpg --output_dir gen_img

(I guess you renamed NST.py to NST_train_single_image.py, right?)
Anyhow, your notes seem to indicate I need to train images on the style images first?
You're notes seem to suggest that there is a lot more involved:
Run this file to train NST on a new style and input pair. At least 2000 epochs are recommended.
Run this file to transfer input image into pre-trained styles. There are 4 pre-trained models are available in models/binaries folder.
To train NST on new style. This will save the pre-trained model which can be used to transfer images afterwards.

To train NST on new style. This will save the pre-trained model which can be used to transfer images afterwards.

### NST_train_single_image.py

This one uses one of the pre-trained models so it's not training here. This is the main file to use.

### stylization_script.py

I have not trained with this one as it requires huge data and more time but just added here in case you want to train your own style. Once you train using this script it saves the pre-trained models that you can later use in stylization_script.py

### training_script.py

You don't have to train unless you want a new style.

## Previous use of the NST script

The script was called like this:

python apps/toonify/src/NST.py

```py
parser.add_argument(
    '--input_dir', default='C:/Users/timof/repos/timofeysie/satisfactory/apps/toonify/src/test_img/Trump-Kim_Summit_June_2018_v4.jpg')
parser.add_argument(
    '--style_image', default='C:/Users/timof/repos/timofeysie/satisfactory/apps/toonify/src/style_img/Overlooking_Stampa_Augusto_Giacometti_(1908).jpg')
...
save_image(generated_image, "gen_" + opt.input_dir)
```

Here is the list of [commits](https://github.com/timofeysie/satisfactory/commits/develop) to figure out how to use the scripts.

### Scripts added

Scripts and commit messages:

```txt
NST_train_single_image.py - NST new pre-trained models and training code
stylization_script.py - NST new pre-trained models and training code
test.py - NST pre-trained models and training code
test_image_cyclegan.py - Cyclegan pre-trained models and test script
training_script.py - NST new pre-trained models and training code
```

In test.py, paths were changed to relative paths:

parser.add_argument('--input_dir', default = './test_img/')

However, the generate images does not seem to be working anymore.

I have seen some strange images in the apps\toonify\src\cartooned_img directory.

Why are these files there?

Alex_De_Minaur_%2828938823637%29_sezanne.jpg
Alex_De_Minaur_%2843875330731%29_sezanne.jpg
Bundesarchiv_Bild_146-1991-014-09%2C_Stein-Pfalz%2C_Ritter_und_Justin_bei_Blutabnahme_sezanne.jpg

I think these must be the failed images El was talking about.  Was that using the Cezanne model.  Since his French name has an accent over the e: Paul CÃ©zanne, I'm guessing that's what that encoded character is: %29_sezanne

When I run the test.py script from the command line, I get this:

```txt
> python apps/toonify/src/test.py --style Hosoda --gpu 0
Traceback (most recent call last):
  File "C:\Users\timof\repos\timofeysie\satisfactory\apps\toonify\src\test.py", line 27, in <module>
    model.load_state_dict(torch.load(os.path.join(opt.model_path, opt.style + '_net_G_float.pth')), strict=False)
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 594, in load   
    with _open_file_like(f, 'rb') as opened_file:
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 230, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 211, in __init__
    super(_open_file, self).__init__(open(name, mode))
FileNotFoundError: [Errno 2] No such file or directory: './pretrained_model/Hosoda_net_G_float.pth'
PS C:\Users\timof\repos\timofeysie\satisfactory>
```

Change the new version:

parser.add_argument('--input_dir', default = './test_img/')

Back to the old version:

parser.add_argument('--input_dir', default='apps/toonify/src/test_img/')

And it works.

## What happened to the download images?

linkWrapper {
  links: 'https://upload.wikimedia.org/wikipedia/commons/c/c9/Corporal_Ben_Roberts-Smith_VC_investiture_%285%29.jpg'
}

Error: ENOENT: no such file or directory, open 'C:\Users\timof\repos\satisfactory\apps\toonify\src\test_img\Corporal_Ben_Roberts-Smith_VC_investiture_%285%29.jpg'

What is this?

Version: typescript 4.2.4
Time: 426ms
Command failed: taskkill /pid 4868 /T /F
ERROR: The process "4868" not found.

Debugger listening on ws://localhost:60881/6b8b907e-b01e-439a-80bf-7b1d06e12e08
Debugger listening on ws://localhost:60882/6b8b907e-b01e-439a-80bf-7b1d06e12e08
For help, see: <https://nodejs.org/en/docs/inspector>
[Nest] 6312   - 05/03/2022, 11:14:49 pm   [NestFactory] Starting Nest application...
[Nest] 6312   - 05/03/2022, 11:14:49 pm   [InstanceLoader] HttpModule dependencies initialized +49ms
...

The linkWrapper {
  links: 'https://upload.wikimedia.org/wikipedia/commons/c/ce/Corporal_Ben_Roberts-Smith_VC_investiture_%287%29.jpg'
}

### The steps for a download

libs\trends\src\lib\containers\trends\trends.component.ts

```ts
preFillForm() {
...
    // download images
    this.downloadImages();
```

### The actual result

```txt
id Milwaukee Bucks
ImagesService.fullUrl <https://commons.wikimedia.org/w/index.php?search=Milwaukee> Bucks&title=Special:MediaSearch&go=Go
list 40
linkWrapper {
  links: 'https://upload.wikimedia.org/wikipedia/commons/0/02/President_Joe_Biden_with_the_Milwaukee_Bucks_%2803%29.jpg'
}
gan.controller.downloadImage: name {
  path: 'https://upload.wikimedia.org/wikipedia/commons/0/02/',
  filename: 'President_Joe_Biden_with_the_Milwaukee_Bucks_%2803%29.jpg'
}
file stream created
gan.controller.downloadImage: axios ref response
[Nest] 8544   - 08/03/2022, 12:46:24 am   [ExceptionsHandler] ENOENT: no such file or directory, open 'C:\Users\timof\repos\satisfactory\apps\toonify\src\test_img\President_Joe_Biden_with_the_Milwaukee_Bucks_%2803%29.jpg' +349800ms
```

Did the download happen and the image just not save?

If we look at the response object, it seems to indicate yes:

gan.controller.downloadImage: axios ref response {
  status: 200,
  statusText: 'OK',
  headers: {
    date: 'Mon, 07 Mar 2022 05:15:42 GMT',

The error is in the reject block in the gan.controller.downloadImage() function which previously worked:

```ts
    return new Promise((resolve, reject) => {
      writer.on('finish', () => this.ganService.kickOffGan());
      writer.on('error', () => { reject }
```

So look at the path used.

C:\Users\timof\repos\satisfactory\apps\toonify\src\test_img\President_Joe_Biden_with_the_Milwaukee_Bucks_%2803%29.jpg' +349800ms

C:\Users\timof\repos\timofeysie\satisfactory-broken\apps\toonify\src\test_img

Right.  The repos position has change because the develop branch lost it's ref and the project was re-cloned and moved up a directory.  Where is there a hardwired path?

There were no directories:

- test_img
- gen_image
- cartooned_img

Someone needs to remember that these directories will have to be created if the project is moved again or cloned by someone else.  Why aren't they in the git repo?  They are not in the gitignore.  As far as is know, the app does not hard-wire the satisfactory directory anywhere

Error: ENOENT: no such file or directory, open 'C:\Users\timof\repos\satisfactory\apps\nest-demo\src\app\bart\summaries\https%3A%2F%2Fwww.fearthesword.com%2F2022%2F3%2F7%2F22964370%2Fevan-mobley-offensive-rebounding-breakdown.txt'

Maybe we need the apps\nest-demo\src\app\bart\summaries directory also?

The post is made, but there is no S3 bucket upload.  Here we go again:

writeBucketFile: File not found ./apps/nest-demo/src/app/gan/bucket/Zydrunas_Ilgauskas_Cleveland_Cavaliers_Hayao.jpg.json

Create that directory, but then we see this error:

bart.service.getArticleSummary: path ./apps/nest-demo/src/app/bart/summaries/https%3A%2F%2Fmeaww.com%2Frobert-kraft-dana-blumberg-engaged-tommy-hilfiger-engaged.txt
gan.service.upload: err [Error: ENOENT: no such file or directory, open 'C:\Users\timof\repos\satisfactory\apps\toonify\src\cartooned_img\Brian_Nord,_Becoming_Interplanetary_Paprika.jpg'] {

A simple solution would be to create the directory if it doesn't exist.  Does FS have such a function?  If not, here is the list of paths that do not exist when a repo is cloned:

### Directories needed

- ./apps/toonify/src/cartooned_img/test_img
- ./apps/toonify/src/cartooned_img/gen_image
- ./apps/toonify/src/cartooned_img/cartooned_img
- ./apps/nest-demo/src/app/bart/summaries
- ./apps/nest-demo/src/app/gan/bucket/

## Back to the generate list function

Now that the download, generate and upload are working again, it's back to this.

apps\nest-demo\src\app\products\products.service.ts

The faulty logic goes something like this:

```ts
fs.readdir('./posts', (err, files) => {
  const articles = [];
  for (const [index, file] of files.entries()) {
    fs.readFile('./posts/' + file, 'utf-8', (err, file) => {
      const jsonFile = JSON.parse(file);
      const article = { ... fill properties ... }
      articles.push(article);
    })
  }
  console.log('articles', articles.length);
})
```

However, we get 0 each time for the articles, because the fs functions are async.

readFileSync could be used, but it has a different signature than the async version.

Trying async/await didn't change anything.

Just doing this in a procedural way works:

```ts
const fileContents = fs.readFileSync('./posts/' + file, 'utf-8');
```

## Cycle Gan

El - Sat, Mar 05: have just pushed cyclegan
there are 4 pre-trained models but only one is generating some reasonable images
will modify the notes tomorrow morning

apps\toonify\src\test_image_cyclegan.py

apps\toonify\src\test_img\Paddy_Pimblett_ufc.jpg

python apps\toonify\src\test_image_cyclegan.py --file apps\toonify\src\test_img\Paddy_Pimblett_ufc.jpg

Or this:

```shell
python apps\toonify\src\test_image_cyclegan.py --file apps\toonify\src\test_img\Paddy_Pimblett_ufc.jpg --model-name apps\toonify\src\cezanne2photo/netG_B2A.pth
Namespace(file='apps\\toonify\\src\\test_img\\Paddy_Pimblett_ufc.jpg', model_name='apps\\toonify\\src\\cezanne2photo/netG_B2A.pth', cuda=False, image_size=256, manualSeed=None)
Random Seed:  6480
Traceback (most recent call last):
  File "C:\Users\timof\repos\satisfactory\apps\toonify\src\test_image_cyclegan.py", line 58, in <module>
    model.load_state_dict(torch.load(args.model_name), strict=False)
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 608, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 787, in_legacy_load
    result = unpickler.load()
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 743, in persistent_load
    deserialized_objects[root_key] = restore_location(obj, location)
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 175, in default_restore_location
    result = fn(storage, location)
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 151, in _cuda_deserialize
    device = validate_cuda_device(location)
  File "C:\Users\timof\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\serialization.py", line 135, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.  
```

Changed this:

```py
# Load state dicts
model.load_state_dict(torch.load(
    args.model_name, map_location=torch.device('cpu')), strict=False)
```

And get this:

```shell
> python apps\toonify\src\test_image_cyclegan.py --file apps\toonify\src\test_img\Paddy_Pimblett_ufc.jpg --model-name apps\toonify\src\cezanne2photo/netG_B2A.pth
Namespace(file='apps\\toonify\\src\\test_img\\Paddy_Pimblett_ufc.jpg', model_name='apps\\toonify\\src\\cezanne2photo/netG_B2A.pth', cuda=False, image_size=256, manualSeed=None)
Random Seed:  8122
cost 0.8493s
```

This should be the output dir:

apps\toonify\src\cartooned_img

apps\toonify\src\pretrained_model\vangogh_net_G_float.pth

python apps\toonify\src\test_image_cyclegan.py --file apps\toonify\src\test_img\Paddy_Pimblett_ufc.jpg --model-name apps\toonify\src\pretrained_model\vangogh_net_G_float.pth

This doesn't do much.  I can see why it was abandoned.

### Using NST_train_single_image.py

Args:

parser.add_argument('--input_dir', default = 'rafa.jpg')
parser.add_argument('--style_image', default = 'mystyle.jpeg')

python apps\toonify\src\NST_train_single_image.py --input_dir apps\toonify\src\cartooned_img\Paddy_Pimblett_ufc_Shinkai.jpg --style_image WLANL_-_Techdiva_1.0_-_Korenveld_onder_onweerslucht_,_Vincent_van_Gogh_(1890).jpg

```shell
detail : The term 'detail' is not recognized as the name of a cmdlet, function, script file, or 
operable program. Check the spelling of the name, or if a path was included, verify that the path is  
correct and try again.
At line:1 char:192
+ ... e WLANL_-_Techdiva_1.0_-_Korenveld_onder_onweerslucht_(detail)_,_Vinc ...
+                                                            ~~~~~~
    + CategoryInfo          : ObjectNotFound: (detail:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

python apps\toonify\src\NST_train_single_image.py --input_dir apps\toonify\src\cartooned_img\Paddy_Pimblett_ufc_Shinkai.jpg --style_image apps\toonify\src\style_img\WLANL_-_Techdiva_1.0_-_Korenveld_onder_onweerslucht_Vincent_van_Gogh_1890.jpg

FileNotFoundError: [Errno 2] No such file or directory: 'gen_apps\\toonify\\src\\cartooned_img\\Paddy_Pimblett_ufc_Shinkai.jpg'

The args are set like this in the NST script:

parser.add_argument('--input_dir', default = 'rafa.jpg')
parser.add_argument('--style_image', default = 'mystyle.jpeg')

save_image(generated_image, "apps/toonify/src/gen_image/")
x 
### What happened to the download images? II

linkWrapper {
  links: 'https://upload.wikimedia.org/wikipedia/commons/c/c9/Corporal_Ben_Roberts-Smith_VC_investiture_%285%29.jpg'
}

Error: ENOENT: no such file or directory, open 'C:\Users\timof\repos\satisfactory\apps\toonify\src\test_img\Corporal_Ben_Roberts-Smith_VC_investiture_%285%29.jpg'

What is this?

Version: typescript 4.2.4
Time: 426ms
Command failed: taskkill /pid 4868 /T /F
ERROR: The process "4868" not found.

Debugger listening on ws://localhost:60881/6b8b907e-b01e-439a-80bf-7b1d06e12e08
Debugger listening on ws://localhost:60882/6b8b907e-b01e-439a-80bf-7b1d06e12e08
For help, see: <https://nodejs.org/en/docs/inspector>
[Nest] 6312   - 05/03/2022, 11:14:49 pm   [NestFactory] Starting Nest application...
[Nest] 6312   - 05/03/2022, 11:14:49 pm   [InstanceLoader] HttpModule dependencies initialized +49ms
...

The linkWrapper {
  links: 'https://upload.wikimedia.org/wikipedia/commons/c/ce/Corporal_Ben_Roberts-Smith_VC_investiture_%287%29.jpg'
}

## Component for the articles list

nx g @nrwl/angular:component components/article-list -project=products

### A google Trend item

```html
<a
  ng-href="https://www.thelcn.com/news/local/winter-storm-warning-for-wyoming-county-rest-of-glow-to-see-winter-weather-advisory/article_da47ac55-1229-54fd-b376-d314ce619f97.html"
  ng-repeat="item in list track by $index"
  ng-include="'/components/feed/feed-item-carousel-directive/feed-item-carousel-atom.html'"
  title="Winter storm warning for Wyoming County, rest of GLOW to see ..."
  target="_blank"
  track="['Trending Searches ' + selectedTab, 'click', 'carousel item: ' + title]"
  ng-style="{ width: atomWidth }"
  href="https://www.thelcn.com/news/local/winter-storm-warning-for-wyoming-county-rest-of-glow-to-see-winter-weather-advisory/article_da47ac55-1229-54fd-b376-d314ce619f97.html"
  style="width: 264px"
  ><div class="feed-item-carousel-item">
    <div class="carousel-image-wrapper">
      <!----><img
        ng-src="https://t2.gstatic.com/images?q=tbn:ANd9GcTmCTU-uIPHYdDIaLQpEsYBJ1lChjmMC1G8Y61SC84CmttgfZRoQ3HoUFMBKn2NOrqJqHQdjeOz"
        ng-if="::item.imgUrl"
        role="presentation"
        src="https://t2.gstatic.com/images?q=tbn:ANd9GcTmCTU-uIPHYdDIaLQpEsYBJ1lChjmMC1G8Y61SC84CmttgfZRoQ3HoUFMBKn2NOrqJqHQdjeOz"
      /><!---->
      <!---->
      <div class="image-text" ng-if="item.shouldShowImageSource">
        The Livingston County News
      </div>
      <!---->
    </div>
    <div
      class="carousel-text-wrapper with-image"
      ng-class="{ 'with-image': item.imgUrl }"
    >
      <div class="item-title" multi-line-ellipsis="">
        Winter storm warning for Wyoming County, rest of ...
      </div>
      <div class="item-subtitles" title="The Livingston County News â¢ 1h ago">
        <span>The Livingston County News</span> â¢ <span>1h ago</span>
      </div>
    </div>
  </div>
</a>
```

## Yet another failed git repo

git branch

Failed to resolve HEAD as a valid ref.

git clone <https://github.com/timofeysie/satisfactory.git>

fatal: unable to access 'https://github.com/timofeysie/satisfactory.git/':
OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to github.com:443

tried
git clone git@github.com:timofeysie/satisfactory.git

Cloning into 'satisfactory'...
The authenticity of host 'github.com (52.64.108.95)' can't be established.
ECDSA key fingerprint is SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? y
Please type 'yes', 'no' or the fingerprint: yes
Warning: Permanently added 'github.com,52.64.108.95' (ECDSA) to the list of known hosts.
Connection reset by 52.64.108.95 port 22
fatal: Could not read from remote repository.

Please make sure you have the correct access rights

Then tried the http one again:

git clone <https://github.com/timofeysie/satisfactory.git>
Cloning into 'satisfactory'...
remote: Enumerating objects: 4375, done.
remote: Counting objects: 100% (151/151), done.
remote: Compressing objects: 100% (95/95), done.
error: RPC failed; curl 56 OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 10054
fatal: the remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed

I'm actually trying to upload a video right now so this seemed to be why the clone didn't work.  I eventually cancelled the upload and tried again and it worked.

After cloning a new repo, and copying the files over, it's all working again.

Cannot find module '@nestjs/config' or its corresponding type declarations.ts(2307)

## On post form error

trends.component.ts:190 Error caught at Subscriber [object Object]

## Todo Raised by Wolves Season 2 Episode 3

### Tooltips to add

- Wikipedia link: Edit the extension to create a link, ie: https://en.wikipedia.org/wiki/ + text in the field below and press 'update'.

- News link: AP News is the preferred link, but add any full url link to be used in the field below and press 'update'.

- types: A single AI type requires only one Type section, the second one can be deleted.  An AI vs Art post requires two Type sections.  A mashup type requires two also.

### Mashup types

- Must be able to show two images that will be downloaded.
- Add link label and link url fields on the post form under the current link fields to edit the AP/other links.

### Random

done - Add "copy to meta-description" button next to "Use generated summary" button.
- Add mashup to keywords for mashup type.  Add AI vs Art, AI generated art for single.
done - Move identical parts of libs\products\src\lib\containers\product-detail\product-detail.component.html into dumb shared components

### Issues

- Choose generated image doesn't appear to be working for mashup type.

### Notes

#### Move identical parts of product-detail.component.html into dumb shared components

nx g @nrwl/angular:component components/detail-links -project=products

'r' is not found in schema

npm install -g nx

Didn't help.  Sound [this SO](https://stackoverflow.com/questions/67833697/nx-not-able-to-generate-component-after-upgrading-to-v12) similar question.

Problem solved: removed the last entry for "@nrwl/angular:component" and it now seems to work!

That might be the angular.json file, but the answer is a one liner so doesn't say.

## Mashup second news link label and url fields

While adding these, saw this error:

core.js:6456 ERROR TypeError: Cannot read properties of undefined (reading 'length')
    at TrendsComponent.getCommonsImageUrl (trends.component.ts:427:55)

Looks like an issue with the extension function.

const ext = this.findExtension(urlPage);

It's a terrible function.  The second image is not getting downloaded either.

We need to show the user the selected image in the card also.  Not the image, but the url.

The findExtension: urlPage <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/04/Logo_Boston_Bruins_1934.gif/147px-Logo_Boston_Bruins_1934.gif" data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/04/Logo_Boston_Bruins_1934.gif/147px-Logo_Boston_Bruins_1934.gif" alt="Logo Boston Bruins 1934.gif" loading="lazy" class="sd-image" style="height: 100% !important; max-width: 445px !important; max-height: 545px;">

It's a gif.  Duh.

So downloadImages.two is undefined

<https://upload.wikimedia.org/wikipedia/commons/0/04/Logo_Boston_Bruins_1934.gif>

Seems to be working now.

We really need scrape text buttons in the trends list and then separate summarize buttons in the post form!

Also, make a mashup of MikelArteta and the stadium to complete that post.
